File: mobile/flutter/lib/data/models/kegel.dart
import 'package:json_annotation/json_annotation.dart';

part 'kegel.g.dart';

// ============================================================================
// ENUMS
// ============================================================================

enum KegelFocusArea {
  @JsonValue('general')
  general,
  @JsonValue('male_specific')
  maleSpecific,
  @JsonValue('female_specific')
  femaleSpecific,
  @JsonValue('postpartum')
  postpartum,
  @JsonValue('prostate_health')
  prostateHealth,
}

extension KegelFocusAreaExtension on KegelFocusArea {
  String get displayName {
    switch (this) {
      case KegelFocusArea.general:
        return 'General';
      case KegelFocusArea.maleSpecific:
        return 'Male-Specific';
      case KegelFocusArea.femaleSpecific:
        return 'Female-Specific';
      case KegelFocusArea.postpartum:
        return 'Postpartum';
      case KegelFocusArea.prostateHealth:
        return 'Prostate Health';
    }
  }
}

enum KegelLevel {
  @JsonValue('beginner')
  beginner,
  @JsonValue('intermediate')
  intermediate,
  @JsonValue('advanced')
  advanced,
}

extension KegelLevelExtension on KegelLevel {
  String get displayName {
    switch (this) {
      case KegelLevel.beginner:
        return 'Beginner';
      case KegelLevel.intermediate:
        return 'Intermediate';
      case KegelLevel.advanced:
        return 'Advanced';
    }
  }
}

enum ReminderFrequency {
  @JsonValue('once')
  once,
  @JsonValue('twice')
  twice,
  @JsonValue('three_times')
  threeTimes,
  @JsonValue('hourly')
  hourly,
}

enum KegelSessionType {
  @JsonValue('quick')
  quick,
  @JsonValue('standard')
  standard,
  @JsonValue('advanced')
  advanced,
  @JsonValue('custom')
  custom,
}

enum KegelPerformedDuring {
  @JsonValue('warmup')
  warmup,
  @JsonValue('cooldown')
  cooldown,
  @JsonValue('standalone')
  standalone,
  @JsonValue('daily_routine')
  dailyRoutine,
  @JsonValue('other')
  other,
}

// ============================================================================
// KEGEL PREFERENCES MODEL
// ============================================================================

@JsonSerializable()
class KegelPreferences {
  final String id;
  @JsonKey(name: 'user_id')
  final String userId;
  @JsonKey(name: 'kegels_enabled')
  final bool kegelsEnabled;
  @JsonKey(name: 'include_in_warmup')
  final bool includeInWarmup;
  @JsonKey(name: 'include_in_cooldown')
  final bool includeInCooldown;
  @JsonKey(name: 'include_as_standalone')
  final bool includeAsStandalone;
  @JsonKey(name: 'include_in_daily_routine')
  final bool includeInDailyRoutine;
  @JsonKey(name: 'daily_reminder_enabled')
  final bool dailyReminderEnabled;
  @JsonKey(name: 'daily_reminder_time')
  final String? dailyReminderTime;
  @JsonKey(name: 'reminder_frequency')
  final ReminderFrequency reminderFrequency;
  @JsonKey(name: 'target_sessions_per_day')
  final int targetSessionsPerDay;
  @JsonKey(name: 'target_duration_seconds')
  final int targetDurationSeconds;
  @JsonKey(name: 'current_level')
  final KegelLevel currentLevel;
  @JsonKey(name: 'focus_area')
  final KegelFocusArea focusArea;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;
  @JsonKey(name: 'updated_at')
  final DateTime updatedAt;

  KegelPreferences({
    required this.id,
    required this.userId,
    this.kegelsEnabled = false,
    this.includeInWarmup = false,
    this.includeInCooldown = false,
    this.includeAsStandalone = false,
    this.includeInDailyRoutine = false,
    this.dailyReminderEnabled = false,
    this.dailyReminderTime,
    this.reminderFrequency = ReminderFrequency.twice,
    this.targetSessionsPerDay = 3,
    this.targetDurationSeconds = 300,
    this.currentLevel = KegelLevel.beginner,
    this.focusArea = KegelFocusArea.general,
    required this.createdAt,
    required this.updatedAt,
  });

  factory KegelPreferences.fromJson(Map<String, dynamic> json) =>
      _$KegelPreferencesFromJson(json);

  Map<String, dynamic> toJson() => _$KegelPreferencesToJson(this);

  KegelPreferences copyWith({
    String? id,
    String? userId,
    bool? kegelsEnabled,
    bool? includeInWarmup,
    bool? includeInCooldown,
    bool? includeAsStandalone,
    bool? includeInDailyRoutine,
    bool? dailyReminderEnabled,
    String? dailyReminderTime,
    ReminderFrequency? reminderFrequency,
    int? targetSessionsPerDay,
    int? targetDurationSeconds,
    KegelLevel? currentLevel,
    KegelFocusArea? focusArea,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return KegelPreferences(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      kegelsEnabled: kegelsEnabled ?? this.kegelsEnabled,
      includeInWarmup: includeInWarmup ?? this.includeInWarmup,
      includeInCooldown: includeInCooldown ?? this.includeInCooldown,
      includeAsStandalone: includeAsStandalone ?? this.includeAsStandalone,
      includeInDailyRoutine:
          includeInDailyRoutine ?? this.includeInDailyRoutine,
      dailyReminderEnabled: dailyReminderEnabled ?? this.dailyReminderEnabled,
      dailyReminderTime: dailyReminderTime ?? this.dailyReminderTime,
      reminderFrequency: reminderFrequency ?? this.reminderFrequency,
      targetSessionsPerDay: targetSessionsPerDay ?? this.targetSessionsPerDay,
      targetDurationSeconds:
          targetDurationSeconds ?? this.targetDurationSeconds,
      currentLevel: currentLevel ?? this.currentLevel,
      focusArea: focusArea ?? this.focusArea,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

// ============================================================================
// KEGEL SESSION MODEL
// ============================================================================

@JsonSerializable()
class KegelSession {
  final String id;
  @JsonKey(name: 'user_id')
  final String userId;
  @JsonKey(name: 'session_date')
  final DateTime sessionDate;
  @JsonKey(name: 'session_time')
  final String? sessionTime;
  @JsonKey(name: 'duration_seconds')
  final int durationSeconds;
  @JsonKey(name: 'reps_completed')
  final int? repsCompleted;
  @JsonKey(name: 'hold_duration_seconds')
  final int? holdDurationSeconds;
  @JsonKey(name: 'session_type')
  final KegelSessionType sessionType;
  @JsonKey(name: 'exercise_name')
  final String? exerciseName;
  @JsonKey(name: 'performed_during')
  final KegelPerformedDuring? performedDuring;
  @JsonKey(name: 'workout_id')
  final String? workoutId;
  @JsonKey(name: 'difficulty_rating')
  final int? difficultyRating;
  final String? notes;
  @JsonKey(name: 'created_at')
  final DateTime createdAt;

  KegelSession({
    required this.id,
    required this.userId,
    required this.sessionDate,
    this.sessionTime,
    required this.durationSeconds,
    this.repsCompleted,
    this.holdDurationSeconds,
    this.sessionType = KegelSessionType.standard,
    this.exerciseName,
    this.performedDuring,
    this.workoutId,
    this.difficultyRating,
    this.notes,
    required this.createdAt,
  });

  factory KegelSession.fromJson(Map<String, dynamic> json) =>
      _$KegelSessionFromJson(json);

  Map<String, dynamic> toJson() => _$KegelSessionToJson(this);
}

// ============================================================================
// KEGEL STATS MODEL
// ============================================================================

@JsonSerializable()
class KegelStats {
  @JsonKey(name: 'user_id')
  final String userId;
  @JsonKey(name: 'kegels_enabled')
  final bool kegelsEnabled;
  @JsonKey(name: 'target_sessions_per_day')
  final int targetSessionsPerDay;
  @JsonKey(name: 'total_days_practiced')
  final int totalDaysPracticed;
  @JsonKey(name: 'total_sessions')
  final int totalSessions;
  @JsonKey(name: 'total_duration_seconds')
  final int totalDurationSeconds;
  @JsonKey(name: 'avg_session_duration')
  final int avgSessionDuration;
  @JsonKey(name: 'sessions_today')
  final int sessionsToday;
  @JsonKey(name: 'sessions_last_7_days')
  final int sessionsLast7Days;
  @JsonKey(name: 'current_streak')
  final int currentStreak;
  @JsonKey(name: 'longest_streak')
  final int longestStreak;
  @JsonKey(name: 'daily_goal_met_today')
  final bool dailyGoalMetToday;

  KegelStats({
    required this.userId,
    this.kegelsEnabled = false,
    this.targetSessionsPerDay = 3,
    this.totalDaysPracticed = 0,
    this.totalSessions = 0,
    this.totalDurationSeconds = 0,
    this.avgSessionDuration = 0,
    this.sessionsToday = 0,
    this.sessionsLast7Days = 0,
    this.currentStreak = 0,
    this.longestStreak = 0,
    this.dailyGoalMetToday = false,
  });

  factory KegelStats.fromJson(Map<String, dynamic> json) =>
      _$KegelStatsFromJson(json);

  Map<String, dynamic> toJson() => _$KegelStatsToJson(this);

  String get totalDurationFormatted {
    final hours = totalDurationSeconds ~/ 3600;
    final minutes = (totalDurationSeconds % 3600) ~/ 60;
    if (hours > 0) {
      return '${hours}h ${minutes}m';
    }
    return '${minutes}m';
  }
}

// ============================================================================
// KEGEL EXERCISE MODEL
// ============================================================================

@JsonSerializable()
class KegelExercise {
  final String id;
  final String name;
  @JsonKey(name: 'display_name')
  final String displayName;
  final String description;
  final List<String> instructions;
  @JsonKey(name: 'target_audience')
  final String targetAudience;
  @JsonKey(name: 'focus_muscles')
  final List<String> focusMuscles;
  final String difficulty;
  @JsonKey(name: 'default_duration_seconds')
  final int defaultDurationSeconds;
  @JsonKey(name: 'default_reps')
  final int defaultReps;
  @JsonKey(name: 'default_hold_seconds')
  final int defaultHoldSeconds;
  @JsonKey(name: 'rest_between_reps_seconds')
  final int restBetweenRepsSeconds;
  final List<String> benefits;
  @JsonKey(name: 'video_url')
  final String? videoUrl;
  @JsonKey(name: 'animation_type')
  final String? animationType;

  KegelExercise({
    required this.id,
    required this.name,
    required this.displayName,
    required this.description,
    this.instructions = const [],
    this.targetAudience = 'all',
    this.focusMuscles = const [],
    this.difficulty = 'beginner',
    this.defaultDurationSeconds = 30,
    this.defaultReps = 10,
    this.defaultHoldSeconds = 5,
    this.restBetweenRepsSeconds = 5,
    this.benefits = const [],
    this.videoUrl,
    this.animationType,
  });

  factory KegelExercise.fromJson(Map<String, dynamic> json) =>
      _$KegelExerciseFromJson(json);

  Map<String, dynamic> toJson() => _$KegelExerciseToJson(this);

  bool get isForMen => targetAudience == 'all' || targetAudience == 'male';
  bool get isForWomen => targetAudience == 'all' || targetAudience == 'female';
}

// ============================================================================
// KEGEL DAILY GOAL MODEL
// ============================================================================

@JsonSerializable()
class KegelDailyGoal {
  @JsonKey(name: 'user_id')
  final String userId;
  final DateTime date;
  @JsonKey(name: 'goal_met')
  final bool goalMet;
  @JsonKey(name: 'sessions_completed')
  final int sessionsCompleted;
  @JsonKey(name: 'target_sessions')
  final int targetSessions;
  final int remaining;

  KegelDailyGoal({
    required this.userId,
    required this.date,
    this.goalMet = false,
    this.sessionsCompleted = 0,
    this.targetSessions = 3,
    this.remaining = 3,
  });

  factory KegelDailyGoal.fromJson(Map<String, dynamic> json) =>
      _$KegelDailyGoalFromJson(json);

  Map<String, dynamic> toJson() => _$KegelDailyGoalToJson(this);

  double get progressPercent =>
      targetSessions > 0 ? (sessionsCompleted / targetSessions).clamp(0.0, 1.0) : 0.0;
}

File: backend/api/v1/kegel.py
"""
Kegel/Pelvic Floor API Endpoints
API routes for kegel preferences, session tracking, and pelvic floor exercises.
"""

from fastapi import APIRouter, HTTPException, Query, Request
from typing import Optional, List
from datetime import date, datetime, timedelta
from uuid import UUID

from models.hormonal_health import (
    KegelPreferences, KegelPreferencesCreate, KegelPreferencesUpdate,
    KegelSession, KegelSessionCreate,
    KegelStats, KegelExercise, KegelDailyGoal,
    KegelFocusArea, KegelLevel
)
from core.supabase_client import get_supabase
from core.timezone_utils import resolve_timezone, get_user_today

router = APIRouter(prefix="/kegel", tags=["Kegel/Pelvic Floor"])


# ============================================================================
# KEGEL PREFERENCES ENDPOINTS
# ============================================================================

@router.get("/preferences/{user_id}", response_model=Optional[KegelPreferences])
async def get_kegel_preferences(user_id: UUID):
    """Get user's kegel preferences."""
    print(f"🔍 [Kegel] Fetching preferences for user {user_id}")

    try:
        supabase = get_supabase().client
        result = supabase.table("kegel_preferences").select("*").eq("user_id", str(user_id)).execute()

        if not result.data:
            print(f"ℹ️ [Kegel] No preferences found for user {user_id}")
            return None

        print(f"✅ [Kegel] Preferences retrieved for user {user_id}")
        return result.data[0]

    except Exception as e:
        print(f"❌ [Kegel] Error fetching preferences: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.put("/preferences/{user_id}", response_model=KegelPreferences)
async def upsert_kegel_preferences(user_id: UUID, preferences: KegelPreferencesUpdate):
    """Create or update user's kegel preferences."""
    print(f"🔍 [Kegel] Upserting preferences for user {user_id}")

    try:
        supabase = get_supabase().client

        # Prepare data, excluding None values
        pref_data = {k: v for k, v in preferences.dict().items() if v is not None}
        pref_data["user_id"] = str(user_id)
        pref_data["updated_at"] = datetime.utcnow().isoformat()

        # Convert enums to strings
        for field in ["current_level", "focus_area", "reminder_frequency"]:
            if field in pref_data and pref_data[field] is not None:
                pref_data[field] = pref_data[field].value if hasattr(pref_data[field], 'value') else pref_data[field]

        # Convert time to string
        if "daily_reminder_time" in pref_data and pref_data["daily_reminder_time"]:
            pref_data["daily_reminder_time"] = pref_data["daily_reminder_time"].isoformat()

        # Check if preferences exist
        existing = supabase.table("kegel_preferences").select("id").eq("user_id", str(user_id)).execute()

        if existing.data:
            # Update existing
            result = supabase.table("kegel_preferences").update(pref_data).eq("user_id", str(user_id)).execute()
        else:
            # Insert new
            pref_data["created_at"] = datetime.utcnow().isoformat()
            result = supabase.table("kegel_preferences").insert(pref_data).execute()

        print(f"✅ [Kegel] Preferences upserted for user {user_id}")
        return result.data[0]

    except Exception as e:
        print(f"❌ [Kegel] Error upserting preferences: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/preferences/{user_id}")
async def delete_kegel_preferences(user_id: UUID):
    """Delete user's kegel preferences."""
    print(f"🔍 [Kegel] Deleting preferences for user {user_id}")

    try:
        supabase = get_supabase().client
        supabase.table("kegel_preferences").delete().eq("user_id", str(user_id)).execute()
        print(f"✅ [Kegel] Preferences deleted for user {user_id}")
        return {"message": "Preferences deleted successfully"}

    except Exception as e:
        print(f"❌ [Kegel] Error deleting preferences: {e}")
        raise HTTPException(status_code=500, detail=str(e))


# ============================================================================
# KEGEL SESSION ENDPOINTS
# ============================================================================

@router.post("/sessions/{user_id}", response_model=KegelSession)
async def create_kegel_session(user_id: UUID, session: KegelSessionCreate):
    """Log a completed kegel session."""
    print(f"🔍 [Kegel] Creating session for user {user_id}")

    try:
        supabase = get_supabase().client

        session_data = session.dict()
        session_data["user_id"] = str(user_id)
        session_data["created_at"] = datetime.utcnow().isoformat()
        session_data["session_time"] = datetime.utcnow().time().isoformat()

        # Convert enums and date
        session_data["session_date"] = session_data["session_date"].isoformat()
        for field in ["session_type", "performed_during"]:
            if session_data.get(field):
                session_data[field] = session_data[field].value if hasattr(session_data[field], 'value') else session_data[field]

        result = supabase.table("kegel_sessions").insert(session_data).execute()

        print(f"✅ [Kegel] Session created for user {user_id}")
        return result.data[0]

    except Exception as e:
        print(f"❌ [Kegel] Error creating session: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/sessions/{user_id}", response_model=List[KegelSession])
async def get_kegel_sessions(
    user_id: UUID,
    start_date: Optional[date] = Query(None),
    end_date: Optional[date] = Query(None),
    limit: int = Query(50, ge=1, le=365)
):
    """Get kegel sessions for a user with optional date range."""
    print(f"🔍 [Kegel] Fetching sessions for user {user_id}")

    try:
        supabase = get_supabase().client

        query = supabase.table("kegel_sessions").select("*").eq("user_id", str(user_id))

        if start_date:
            query = query.gte("session_date", start_date.isoformat())
        if end_date:
            query = query.lte("session_date", end_date.isoformat())

        result = query.order("session_date", desc=True).order("session_time", desc=True).limit(limit).execute()

        print(f"✅ [Kegel] Retrieved {len(result.data)} sessions for user {user_id}")
        return result.data

    except Exception as e:
        print(f"❌ [Kegel] Error fetching sessions: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/sessions/{user_id}/today", response_model=List[KegelSession])
async def get_today_kegel_sessions(user_id: UUID, request: Request):
    """Get today's kegel sessions."""
    print(f"🔍 [Kegel] Fetching today's sessions for user {user_id}")

    try:
        supabase = get_supabase().client
        user_tz = resolve_timezone(request, None, str(user_id))
        today_str = get_user_today(user_tz)
        result = supabase.table("kegel_sessions").select("*").eq(
            "user_id", str(user_id)
        ).eq("session_date", today_str).order("session_time", desc=True).execute()

        return result.data

    except Exception as e:
        print(f"❌ [Kegel] Error fetching today's sessions: {e}")
        raise HTTPException(status_code=500, detail=str(e))


# ============================================================================
# KEGEL STATISTICS ENDPOINTS
# ============================================================================

@router.get("/stats/{user_id}", response_model=KegelStats)
async def get_kegel_stats(user_id: UUID, request: Request):
    """Get kegel exercise statistics for a user."""
    print(f"🔍 [Kegel] Calculating stats for user {user_id}")

    try:
        supabase = get_supabase().client

        # Get preferences
        prefs_result = supabase.table("kegel_preferences").select("*").eq("user_id", str(user_id)).execute()
        prefs = prefs_result.data[0] if prefs_result.data else {"kegels_enabled": False, "target_sessions_per_day": 3}

        # Get all sessions
        sessions_result = supabase.table("kegel_sessions").select("*").eq("user_id", str(user_id)).execute()
        sessions = sessions_result.data

        # Calculate stats
        total_sessions = len(sessions)
        total_duration = sum(s.get("duration_seconds", 0) for s in sessions)
        avg_duration = total_duration // total_sessions if total_sessions > 0 else 0

        # Count unique days
        session_dates = set(s.get("session_date") for s in sessions)
        total_days = len(session_dates)

        # Today's sessions
        user_tz = resolve_timezone(request, None, str(user_id))
        today_str = get_user_today(user_tz)
        today_date = date.fromisoformat(today_str)
        sessions_today = len([s for s in sessions if s.get("session_date") == today_str])

        # Last 7 days
        week_ago = (today_date - timedelta(days=7)).isoformat()
        sessions_last_7 = len([s for s in sessions if s.get("session_date", "") >= week_ago])

        # Calculate streak
        current_streak = 0
        longest_streak = 0
        if session_dates:
            sorted_dates = sorted([date.fromisoformat(d) for d in session_dates], reverse=True)

            # Current streak (consecutive days including today or yesterday)
            check_date = today_date
            if check_date not in [date.fromisoformat(d) for d in session_dates]:
                check_date = today_date - timedelta(days=1)

            temp_streak = 0
            for d in sorted_dates:
                if d == check_date:
                    temp_streak += 1
                    check_date -= timedelta(days=1)
                elif d < check_date:
                    break
            current_streak = temp_streak

            # Longest streak
            temp_streak = 1
            sorted_dates_asc = sorted([date.fromisoformat(d) for d in session_dates])
            for i in range(1, len(sorted_dates_asc)):
                if sorted_dates_asc[i] - sorted_dates_asc[i-1] == timedelta(days=1):
                    temp_streak += 1
                else:
                    longest_streak = max(longest_streak, temp_streak)
                    temp_streak = 1
            longest_streak = max(longest_streak, temp_streak)

        # Check if daily goal met
        target = prefs.get("target_sessions_per_day", 3)
        daily_goal_met = sessions_today >= target

        return KegelStats(
            user_id=str(user_id),
            kegels_enabled=prefs.get("kegels_enabled", False),
            target_sessions_per_day=target,
            total_days_practiced=total_days,
            total_sessions=total_sessions,
            total_duration_seconds=total_duration,
            avg_session_duration=avg_duration,
            sessions_today=sessions_today,
            sessions_last_7_days=sessions_last_7,
            current_streak=current_streak,
            longest_streak=longest_streak,
            daily_goal_met_today=daily_goal_met
        )

    except Exception as e:
        print(f"❌ [Kegel] Error calculating stats: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/daily-goal/{user_id}", response_model=KegelDailyGoal)
async def check_daily_goal(user_id: UUID, request: Request, check_date: date = Query(default=None)):
    """Check if user has met their daily kegel goal."""
    print(f"🔍 [Kegel] Checking daily goal for user {user_id}")

    try:
        supabase = get_supabase().client

        if check_date is None:
            user_tz = resolve_timezone(request, None, str(user_id))
            target_date = date.fromisoformat(get_user_today(user_tz))
        else:
            target_date = check_date

        # Get preferences
        prefs_result = supabase.table("kegel_preferences").select("target_sessions_per_day").eq("user_id", str(user_id)).execute()
        target = prefs_result.data[0].get("target_sessions_per_day", 3) if prefs_result.data else 3

        # Count sessions for the date
        sessions_result = supabase.table("kegel_sessions").select("id").eq(
            "user_id", str(user_id)
        ).eq("session_date", target_date.isoformat()).execute()

        completed = len(sessions_result.data)
        remaining = max(0, target - completed)

        return KegelDailyGoal(
            user_id=str(user_id),
            date=target_date,
            goal_met=completed >= target,
            sessions_completed=completed,
            target_sessions=target,
            remaining=remaining
        )

    except Exception as e:
        print(f"❌ [Kegel] Error checking daily goal: {e}")
        raise HTTPException(status_code=500, detail=str(e))


# ============================================================================
# KEGEL EXERCISES REFERENCE ENDPOINTS
# ============================================================================

@router.get("/exercises", response_model=List[KegelExercise])
async def get_kegel_exercises(
    target_audience: Optional[str] = Query(None, description="Filter by 'all', 'male', or 'female'"),
    difficulty: Optional[KegelLevel] = Query(None),
    focus_area: Optional[KegelFocusArea] = Query(None)
):
    """Get list of pelvic floor exercises."""
    print(f"🔍 [Kegel] Fetching exercises")

    try:
        supabase = get_supabase().client

        query = supabase.table("kegel_exercises").select("*").eq("is_active", True)

        if target_audience:
            # Include exercises for 'all' plus the specific audience
            query = query.in_("target_audience", ["all", target_audience])

        if difficulty:
            query = query.eq("difficulty", difficulty.value)

        result = query.order("sort_order").execute()

        # Filter by focus area if needed (done in Python since it's not a direct column)
        exercises = result.data
        if focus_area:
            focus_area_map = {
                KegelFocusArea.POSTPARTUM: "female",
                KegelFocusArea.PROSTATE_HEALTH: "male",
                KegelFocusArea.MALE_SPECIFIC: "male",
                KegelFocusArea.FEMALE_SPECIFIC: "female",
            }
            if focus_area in focus_area_map:
                target = focus_area_map[focus_area]
                exercises = [e for e in exercises if e.get("target_audience") in ["all", target]]

        print(f"✅ [Kegel] Retrieved {len(exercises)} exercises")
        return exercises

    except Exception as e:
        print(f"❌ [Kegel] Error fetching exercises: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/exercises/{exercise_id}", response_model=KegelExercise)
async def get_kegel_exercise(exercise_id: UUID):
    """Get a specific kegel exercise by ID."""
    print(f"🔍 [Kegel] Fetching exercise {exercise_id}")

    try:
        supabase = get_supabase().client
        result = supabase.table("kegel_exercises").select("*").eq("id", str(exercise_id)).execute()

        if not result.data:
            raise HTTPException(status_code=404, detail="Exercise not found")

        return result.data[0]

    except HTTPException:
        raise
    except Exception as e:
        print(f"❌ [Kegel] Error fetching exercise: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/exercises/by-name/{name}", response_model=KegelExercise)
async def get_kegel_exercise_by_name(name: str):
    """Get a kegel exercise by name."""
    print(f"🔍 [Kegel] Fetching exercise by name: {name}")

    try:
        supabase = get_supabase().client
        result = supabase.table("kegel_exercises").select("*").eq("name", name).execute()

        if not result.data:
            raise HTTPException(status_code=404, detail="Exercise not found")

        return result.data[0]

    except HTTPException:
        raise
    except Exception as e:
        print(f"❌ [Kegel] Error fetching exercise: {e}")
        raise HTTPException(status_code=500, detail=str(e))


# ============================================================================
# WORKOUT INTEGRATION ENDPOINTS
# ============================================================================

@router.get("/for-workout/{user_id}")
async def get_kegels_for_workout(
    user_id: UUID,
    placement: str = Query(..., description="'warmup', 'cooldown', or 'standalone'")
):
    """Get kegel exercises to include in a workout based on user preferences."""
    print(f"🔍 [Kegel] Getting kegels for {placement} for user {user_id}")

    try:
        supabase = get_supabase().client

        # Check preferences
        prefs_result = supabase.table("kegel_preferences").select("*").eq("user_id", str(user_id)).execute()

        if not prefs_result.data:
            return {"include_kegels": False, "exercises": [], "reason": "No kegel preferences set"}

        prefs = prefs_result.data[0]

        if not prefs.get("kegels_enabled"):
            return {"include_kegels": False, "exercises": [], "reason": "Kegels disabled"}

        # Check if this placement is enabled
        placement_field = f"include_in_{placement}"
        if placement == "standalone":
            placement_field = "include_as_standalone"

        if not prefs.get(placement_field):
            return {"include_kegels": False, "exercises": [], "reason": f"Kegels not enabled for {placement}"}

        # Get appropriate exercises
        level = prefs.get("current_level", "beginner")
        focus = prefs.get("focus_area", "general")

        # Map focus area to target audience
        target_audience = "all"
        if focus in ["male_specific", "prostate_health"]:
            target_audience = "male"
        elif focus in ["female_specific", "postpartum"]:
            target_audience = "female"

        # Fetch exercises
        query = supabase.table("kegel_exercises").select("*").eq("is_active", True).eq("difficulty", level)
        if target_audience != "all":
            query = query.in_("target_audience", ["all", target_audience])
        else:
            query = query.eq("target_audience", "all")

        result = query.order("sort_order").limit(3).execute()

        return {
            "include_kegels": True,
            "placement": placement,
            "exercises": result.data,
            "total_duration_seconds": sum(e.get("default_duration_seconds", 30) for e in result.data)
        }

    except Exception as e:
        print(f"❌ [Kegel] Error getting kegels for workout: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/log-from-workout/{user_id}")
async def log_kegels_from_workout(
    user_id: UUID,
    request: Request,
    workout_id: UUID,
    placement: str = Query(..., description="'warmup' or 'cooldown'"),
    duration_seconds: int = Query(..., ge=1),
    exercises_completed: List[str] = Query(default=[])
):
    """Log kegel exercises completed as part of a workout."""
    print(f"🔍 [Kegel] Logging kegels from workout {workout_id} for user {user_id}")

    try:
        supabase = get_supabase().client
        user_tz = resolve_timezone(request, None, str(user_id))

        session_data = {
            "user_id": str(user_id),
            "session_date": get_user_today(user_tz),
            "session_time": datetime.utcnow().time().isoformat(),
            "duration_seconds": duration_seconds,
            "session_type": "standard",
            "performed_during": placement,
            "workout_id": str(workout_id),
            "exercise_name": ", ".join(exercises_completed) if exercises_completed else None,
            "created_at": datetime.utcnow().isoformat()
        }

        result = supabase.table("kegel_sessions").insert(session_data).execute()

        print(f"✅ [Kegel] Logged kegels from workout for user {user_id}")
        return result.data[0]

    except Exception as e:
        print(f"❌ [Kegel] Error logging from workout: {e}")
        raise HTTPException(status_code=500, detail=str(e))
